### PR TITLE
Product CRUD

### PR DESCRIPTION
**Pull request by UtsavSomaiya**
- [ ] In this PR we can perform category CRUD

### CODE REVIEW

File: app/Http/Controllers/Admin/ProductController.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    public function add()
+    {
+        $categories = Category::all();
+
+        return view('admin.products.add', compact('categories'));
+    }
+    public function store(Request $request)
+    {
+        $validatedData = $request->validate([
+            'name' => ['required','min:3','max:255','unique:products,name'],
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/ProductController.php:29]: min:3 not needed

File: app/Http/Controllers/Admin/ProductController.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        $request->file('image')->storeAs('public/image', $name);
+
+        $validatedData['image'] = $name;
+
+        $validatedData['path'] = asset('storage/image').'/'.$name;
+
+        Product::create($validatedData);
+
+        return to_route('products')->with([
+            'success' => 'Products added successfully.'
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/ProductController.php:48]:
```suggestion
            'success' => 'Product added successfully.'
```

File: app/Http/Controllers/Admin/ProductController.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        ]);
+    }
+
+
+    public function edit(Product $product)
+    {
+        $categories = Category::findOrFail($product->category_id)->get();
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/ProductController.php:55]: why used `->get()` is there any reason?

File: app/Http/Controllers/Admin/ProductController.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            ],
+            'price' => ['required'],
+            'category_id' => ['required'],
+            'tax' => ['required'],
+            'stock' => ['required'],
+            'image' => ['image'],
+        ]);
+
+        if ($request->file('image') != null) {
+            Storage::delete('public/image'.'/'.$product->image);
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/ProductController.php:77]: delete after storing a file

File: app/Models/Product.php [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Foundation\Auth\User as Authenticatable;
+
+class Product extends Authenticatable
+{
+    use HasFactory;
+
+    protected $fillable = ['name','price','category_id','tax','stock','image','path'];
```
Comment (hirenkeraliya) [inline app/Models/Product.php:12]: always space after the `,` sign

File: database/migrations/2022_05_25_110122_create_products_table.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('products', function (Blueprint $table) {
+            $table->id();
+            $table->string('name')->unique();
+            $table->float('price');
```
Comment (hirenkeraliya) [inline database/migrations/2022_05_25_110122_create_products_table.php:18]: use decimal

File: database/migrations/2022_05_25_110122_create_products_table.php [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            $table->unsignedBigInteger('category_id');
+            $table->foreign('category_id')->references('id')->on('categories');
```
Comment (hirenkeraliya) [inline database/migrations/2022_05_25_110122_create_products_table.php:20]:
```suggestion
            $table->foreignId('category_id')->constrained();
```

File: database/migrations/2022_05_25_110122_create_products_table.php [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            $table->float('tax');
```
Comment (hirenkeraliya) [inline database/migrations/2022_05_25_110122_create_products_table.php:21]: why tax? as per your answer of my question `Tax is Cart Wide`

File: public/css/admin/custom.css [hunk 9]
```diff
@@ -7,3 +7,14 @@ li:hover
     background-color: gray;
     border-radius: 20px;
 }
+input::-webkit-outer-spin-button,
+input::-webkit-inner-spin-button {
+    -webkit-appearance: none;
+    margin: 0;
+}
+img{
```
Comment (hirenkeraliya) [inline public/css/admin/custom.css:15]: proper indentation

File: public/js/custom.js [hunk 10]
```diff
@@ -6,4 +6,11 @@ function deleteConfirm(type,id) {
             return;
         });
     }
+    if (type == 'products') {
+        alertify.confirm('Delete Product', 'Are you sure to delete this product?', function () {
+            document.getElementById("delete-product-"+id).submit();
+        }, function () {
+            return;
+        });
```
Comment (hirenkeraliya) [inline public/js/custom.js:14]: make this dynamic

File: resources/views/admin/products/add.blade.php [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+            <div class="card-title">
+                <div class="mt-2">
+                    <span class="h4 ms-3 me-5">Add new Product</span>
+                </div>
+            </div>
+            <div class="card-body">
+                <form class="forms-sample"
+                    method="post"
+                    enctype="multipart/form-data"
+                    @empty($product) action="{{ route('products.store') }}" @endempty
```
Comment (hirenkeraliya) [inline resources/views/admin/products/add.blade.php:14]: what if user come for the edit product

File: resources/views/admin/products/add.blade.php [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+                        <label class="pb-1">Tax of the Product</label>
+                        <select class="form-control @error('tax') is-invalid @enderror"
+                            name="tax"
+                            required
+                        >
+                            <option value="">--Select Tax--</option>
+                            <option value="5"
+                                @isset($product)
+                                    @selected($product->tax == "5")
+                                @else
```
Comment (hirenkeraliya) [inline resources/views/admin/products/add.blade.php:79]: make this dynamic

File: resources/views/admin/products/index.blade.php [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+                            <th>Price</th>
+                            <th>Category</th>
+                            <th>Tax</th>
+                            <th>Stock</th>
+                            <th>Image</th>
+                            <th>Action</th>
+                        </tr>
+                    </thead>
+                    <tbody>
+                        @forelse($products as $key => $product)
```
Comment (hirenkeraliya) [inline resources/views/admin/products/index.blade.php:32]: what use of `$key`

**Review by hirenkeraliya — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Admin/ProductController.php [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        $products = Product::with('category')->get();
+
+        return view('admin.products.index', compact('products'));
+    }
+    public function add()
+    {
+        $categories = Category::all();
+
+        return view('admin.products.add', compact('categories'));
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/ProductController.php:24]: rename add to form

**Review by hirenkeraliya — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Admin/ProductController.php (hunk 4 above)
Comment (UtsavSomaiya) [reply app/Http/Controllers/Admin/ProductController.php:77]: > delete after storing a file

delete after updating a file

File: app/Http/Controllers/Admin/ProductController.php [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+                'required',
+                'max:255',
+                Rule::unique('products', 'name')->ignore($product->id),
+            ],
+            'price' => ['required','numeric'],
+            'category_id' => ['required','exists:categories,id'],
+            'tax' => ['required','numeric'],
+            'stock' => ['required','integer'],
+            'image' => ['image'],
+        ]);
```
Comment (hirenkeraliya) [inline app/Http/Controllers/Admin/ProductController.php:75]: duplicate code

File: database/migrations/2022_05_25_110122_create_products_table.php [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('products', function (Blueprint $table) {
+            $table->id();
+            $table->string('name')->unique();
+            $table->decimal('price');
```
Comment (hirenkeraliya) [inline database/migrations/2022_05_25_110122_create_products_table.php:18]:
```suggestion
            $table->decimal('price', 10, 2);
```

File: database/migrations/2022_05_25_110122_create_products_table.php [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+            $table->foreignId('category_id')->constrained();
+            $table->decimal('tax');
```
Comment (hirenkeraliya) [inline database/migrations/2022_05_25_110122_create_products_table.php:20]:
```suggestion
            $table->decimal('tax', 5, 2);
```

File: public/js/custom.js [hunk 18]
```diff
@@ -1,9 +1,7 @@
 function deleteConfirm(type,id) {
-    if (type == "categories") {
-        alertify.confirm('Delete Category', 'Are you sure to delete this category?', function () {
-            document.getElementById("delete-category-"+id).submit();
-        }, function () {
-            return;
-        });
-    }
+    alertify.confirm('Delete  '+ type, 'Are you sure to delete this ' + type + '?', function () {
```
Comment (hirenkeraliya) [inline public/js/custom.js:2]:
```suggestion
    alertify.confirm('Delete  '+ type, 'Are you sure you want to delete this ' + type + '?', function () {
```

**Review by hirenkeraliya — CHANGES_REQUESTED** [review]

**Comment by hirenkeraliya** [comment]
I have checked the 65% percent PR and you pushed the new code of the export in the same pr. What should I do now?

**Pull request closed without merging**